client/order/orderbook.go

```diff
@@ -0,0 +1,365 @@
+package order
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/decred/dcrdex/dex/msgjson"
+	"github.com/decred/dcrdex/dex/order"
+)
+
+var (
+	// defaultQueueCapacity represents the default capacity of
+	// the order note queue.
+	defaultQueueCapacity = 10
+)
+
+// Order represents an ask or bid.
+type Order struct {
+	OrderID  order.OrderID
+	Side     uint8
+	Quantity uint64
+	Rate     uint64
+}
+
+// RemoteOrderBook defines the functions a client tracked order book
+// must implement.
+type RemoteOrderBook interface {
+	// Sync instantiates a client tracked order book with the
+	// current order book snapshot.
+	Sync(*msgjson.OrderBook)
+	// Book adds a new order to the order book.
+	Book(*msgjson.BookOrderNote)
+	// Unbook removes an order from the order book.
+	Unbook(*msgjson.UnbookOrderNote) error
+}
+
+// CachedOrderNote represents a cached order not entry.
+type cachedOrderNote struct {
+	Route     string
+	OrderNote interface{}
+}
+
+// OrderBook represents a client tracked order book.
+type OrderBook struct {
+	seq          uint64
+	marketID     string
+	noteQueue    []*cachedOrderNote
+	noteQueueMtx sync.Mutex
+	orders       map[order.OrderID]*Order
+	ordersMtx    sync.Mutex
+	buys         *orderGroupDepth
+	sells        *orderGroupDepth
+	synced       bool
+	syncedMtx    sync.Mutex
+}
+
+// NewOrderBook creates a new order book.
+func NewOrderBook() *OrderBook {
+	ob := &OrderBook{
+		noteQueue: make([]*cachedOrderNote, 0, defaultQueueCapacity),
+	}
+	return ob
+}
+
+// setSynced sets the synced state of the order book.
+func (ob *OrderBook) setSynced(value bool) {
+	ob.syncedMtx.Lock()
+	ob.synced = value
+	ob.syncedMtx.Unlock()
+}
+
+// isSynced returns the synced state of the order book.
+func (ob *OrderBook) isSynced() bool {
+	ob.syncedMtx.Lock()
+	defer ob.syncedMtx.Unlock()
+	return ob.synced
+}
+
+// cacheOrderNote caches an order note.
+func (ob *OrderBook) cacheOrderNote(route string, entry interface{}) error {
+	note := new(cachedOrderNote)
+
+	switch route {
+	case msgjson.BookOrderRoute:
+		_, ok := entry.(*msgjson.BookOrderNote)
+		if !ok {
+			return fmt.Errorf("failed to cast entry as BookOrderNote")
+		}
+
+		note.Route = route
+		note.OrderNote = entry
+
+	case msgjson.UnbookOrderRoute:
+		_, ok := entry.(*msgjson.UnbookOrderNote)
+		if !ok {
+			return fmt.Errorf("failed to cast entry as UnbookOrderNote")
+		}
+
+		note.Route = route
+		note.OrderNote = entry
```

Comment:
These two lines in both cases can come after/outside the switch since `entry`'s asserted underlying type is not used.

Reply:
handled another issue related to this by simplifying to this:

```
switch route {
case msgjson.BookOrderRoute, msgjson.UnbookOrderRoute:
	note.Route = route
	note.OrderNote = entry

	ob.noteQueueMtx.Lock()
	ob.noteQueue = append(ob.noteQueue, note)
	ob.noteQueueMtx.Unlock()

	return nil

default:
	return fmt.Errorf("unknown route provided %s", route)
}
```